### DISCUSSION TITLE
How far do we want to create `components`

### DISCUSSION THREAD

**Discussion by cavasinf** [kevinpapst/TablerBundle#116] [Q&A]
Hey @kevinpapst,

Just wondering, 
do we want to create simple components only (card, modal, timeline).
Or can we also made them with Extra/Demo from Tabler?

Example with [Setting extra](https://preview.tabler.io/settings.html).
I've created on my side a template with this render but with bootstrap tab version (JS navigation) AND literal hard links (URL navigation).

![Settings extra demo JS](https://user-images.githubusercontent.com/25293190/192234137-a0900b03-f9e2-4da8-99da-e8b82ad60e54.gif)

Do we implement that in this bundle ?

**Comment by kevinpapst** [comment]
It's a good question how far we want to go with the abstraction. Obviously we would all benefit from such a component, as it would reduce the amount of extra HTML needed, which I am trying to avoid as much as possible outside roof macros and embeds.

The problem with such a component is that you have to find a way to move the tab contents between the use of the macros, probably something like stacked embeds. Not sure how to do that.

Edit:
I have a settings page that behaves differently, there is a list on the left side, with scroll anchors, all settings are grouped in cards one below the other. It makes the page much longer and looks a bit overwhelming, but on the other hand it allows to use the CMD+F search inside the browser. 
In other words: I'd like to see more components without doubt, but I don't have an urgent need for it. But if you want to work on it: awesome 😄 I am sure I will use it at some point!

**Reply by cavasinf** [reply]
> The problem with such a component is that you have to find a way to move the tab contents between the use of the macros, probably something like stacked embeds. Not sure how to do that.

I think I maybe understand what you are saying..
You mean that if we need to write long/complex twig for those tab contents, it will be hard if we need to pass those inside a "macro parameter" ?
If it's the case, I've the same problem currently with the accordion component => content is a complex HTML/twig, but macro only waits for parameters.

**Reply by kevinpapst** [reply]
Yes, that's what I am trying to say 😁 Actually it would be good to have multi macros, a big one like the existing that takes everything as parameters, but also helper macros for "start accordion", "start according card", "end accordion card", "end accordion". You could always use 
```
{% set accordion_content %}
long html
{% endset %}
```
And then pass the `accordion_content` variable to the macro... but that feels wrong.

**Reply by cavasinf** [reply]
> ```
> {% set accordion_content %}
> long html
> {% endset %}
> ```
> 
> 
>     
>   
> 
> And then pass the `accordion_content` variable to the macro... but that feels wrong.

That's what I'm doing with the `accordion` component 😭

**Comment by cavasinf** [comment]
> I have a settings page that behaves differently, there is a list on the left side, with scroll anchors, all settings are grouped in cards one below the other. It makes the page much longer and looks a bit overwhelming, but on the other hand it allows to use the CMD+F search inside the browser.

Can you show me that pls ?

**Reply by kevinpapst** [reply]
Login with `susan_super` and `kitten` to https://demo-branch.kimai.org/fr/admin/system-config/

**Reply by cavasinf** [reply]
Yeah OK.
Hard to say what is the best way.
On your settings page, does submitting 1 of the forms, triggers submit of ALL others forms?
Or do you need to "submit" every form one by one?

**Reply by cavasinf** [reply]
Gitlab does the same way as yours, but they have made the "card" collapsed by default.
BUT with a short description text that tells you what you can find in that card SO you can also CMD + F to quickly find what you are looking for without expanding all cards.

Example with CMD + F `Merge checks`:

![Capture d’écran 2022-09-26 à 11 26 53](https://user-images.githubusercontent.com/25293190/192242933-a973571c-4aa8-452f-be98-1f154d358708.png)

![Capture d’écran 2022-09-26 à 11 27 02](https://user-images.githubusercontent.com/25293190/192242938-27ce7699-b0f7-4d47-89e3-563add21dd7c.png)

I find it nicely done like that.

**Reply by kevinpapst** [reply]
You need to submit every form by itself. In 100% of all cases that I observed in screen shares, users edited at most 1-3 settings at a time and saved those changes directly. Not perfectly elegant, but it keeps the save button near to the setting. 
Gitlab UI looks really good, maybe I will steal that idea in the future 😁 
I find it hard to say what works better for end-users, as they tend to behave much different from what I expect as developer...  I like the look, but on the other hand you have to expand all boxes to be able to find the setting that might belong to different areas.
